Improve wxListCtrl in MSW darkmode

MaartenBent opened the pull request:
Draw the background color outside the list-items area on all Windows version, both below and behind the items, and when the list is empty.

Set the text foreground and background colors when custom drawing. So the disabled list will use the correct colors in list or icon view.

Fixes #25990

vadz commented on src/msw/listctrl.cpp:
```diff
@@ -3471,16 +3476,16 @@ void wxListCtrl::OnPaint(wxPaintEvent& event)
 
     // Check if we need to do anything ourselves: either draw the rules or, in
-    // case of using dark mode under Windows 11, erase the unwanted separator
+    // case of using dark mode under Windows, erase the unwanted separator
```
Very minor, but it doesn't make sense to say "under Windows" in this Windows-only code:

```suggestion
    // case of using dark mode, erase the unwanted separator
```

vadz reviewed (approved):
I didn't test it yet, but this looks good, so I'll try to test it and merge it soon, thank you!

MaartenBent committed:
Improve disabled wxListCtrl in MSW dark mode

Draw the background color outside the list-items area on all Windows version,
both below and behind the items, and when the list is empty.

Set the text foreground and background colors when custom drawing. So the disabled
list will use the correct colors in list or icon view.

Fixes #25990

MaartenBent committed:
Improve drawing wxListCtrl gird lines in dark mode

The grid lines are erased when custom drawing the sub-items, so the grid lines need to be redrawn after drawing the item.
OnPaint() is not called after every custom draw (e.g. just hovering over a row will cause a custom draw but no paint),
so it has to be drawn in the custom paint handler.

Add a common function for drawing grid lines and reuse it on OnPaint().

MaartenBent committed:
Fix incorrect wxListCtrl::GetSubItemRect(n, 0)

This was broken in bde6b7a8b8.
When the list is scrolled the x-coordinate will be negative.
So don't start counting from 0, but start from the original x coordinate of the full rectangle.

MaartenBent committed:
Correctly restore dark mode header in MSW wxListCtrl

When removing custom header attributes, the default dark mode attributes need to be restored.

MaartenBent committed:
Fix MSW wxListCtrl custom colours in light mode

This was broken in 3842352b21.

MaartenBent committed:
Improve drawing MSW wxListCtrl focus rect in dark mode

Subtract 2px so the right border is visible.

MaartenBent committed:
Draw hot tracking highlight in MSW wxListCtrl in dark mode

When the mouse is over an item, change the colour of it.

This only works when using Explorer theme, DarkMode_Explorer does not get custom draw calls to remove the colour.

Fixes #26001

MaartenBent committed:
Fix highlighting first item in MSW wxListCtrl in dark mode

The item was also highlighted when hovering over the header because ListView_HitTest
returns true when hovering over the header.
Replace ListView_HitTest by checking if the mouse pointer is in the item rectangle.

MaartenBent committed:
Slightly increase text padding in MSW wxListCtrl in dark mode

Try to get it to look like light mode.
See #26001

MaartenBent committed:
Better fix for MSW wxListCtrl focus rect visibility

When erasing the background, draw outside the item rectangle.

MaartenBent committed:
Try to fix missing MSW wxListCtrl attribute colour in dark mode

See #26001

MaartenBent committed:
Improve MSW wxListCtrl background attribute in dark mode

It extends the full column, including the padding.
To match light mode, there should be 1px between columns that do not have the attribute colour.
The first (most left) column, and columns with checkboxes or images should still use the padding.
